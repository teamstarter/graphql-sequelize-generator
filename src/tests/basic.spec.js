const request = require('supertest')
const { deleteTables } = require('./testDatabase.js')
const { createServer, closeServer, resetDb } = require('./setupTestServer')

/**
 * Starting the tests
 */
describe('Test the API queries', () => {
  let server = null

  beforeAll(async () => {
    server = await createServer()
  })

  afterAll(() => closeServer(server))

  beforeEach(async () => {
    await resetDb()
  })

  afterEach(async () => {
    await deleteTables()
  })

  it('Check that you can query a model and an association', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company {
              id
              name
              users {
                id
                name
              }
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot('All companies')
  })

  it('Check that you can query sub associations', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company {
              id
              name
              users {
                id
                name
                department {
                  id
                }
              }
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot(
      'All companies with users and their department'
    )
  })

  it('Check that you can sort', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company(order: "name,id") {
              id
              name
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot('All companies sorted by name')
  })

  it('Check that you can sort by virtual fields in any order', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company(order: "reverse:userCount,name") {
              id
              userCount
              name
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot(
      'All companies sorted by userCount desc and name asc.'
    )
  })

  it('Sorting can ignore spacing typoes', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company(order: "userCount, name") {
              id
              userCount
              name
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot(
      'All companies sorted by userCount desc and name asc.'
    )
  })

  it('Sorting work on secondary field and associated fields', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            companies: company(order :"type.id,reverse:name") {
              id
              name
              userCount
              type {
                id
              }
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    const companies = response.body.data.companies
    expect(companies).toMatchSnapshot(
      'All companies sorted by userCount desc and name asc.'
    )
  })

  it('One can exclude an associated model', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getDepartments {
            department {
              id
              name
              company {
                id
              }
            }
          }
          &operationName=getDepartments`
      )
      .set('userId', 1)
    expect(response.body.errors).not.toBeUndefined()
    expect(response.body.errors).toMatchSnapshot()
  })

  it('One can exclude a basic field from a model', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getDepartments {
            department {
              id
              name
              updatedAt
            }
          }
          &operationName=getDepartments`
      )
      .set('userId', 1)
    expect(response.body.errors).not.toBeUndefined()
    expect(response.body.errors).toMatchSnapshot()
  })

  it('One can exclude a model from the root of the server.', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getCompanySettings {
            companySetting {
              whiteLabelEnabled
            }
          }
          &operationName=getCompanySettings`
      )
      .set('userId', 1)
    expect(response.body.errors).not.toBeUndefined()
    expect(response.body.errors).toMatchSnapshot()

    const responseSuccess = await request(server)
      .get(
        `/graphql?query=
          query getCompanies {
            company {
              id
              settings {
                whiteLabelEnabled
              }
            }
          }
          &operationName=getCompanies`
      )
      .set('userId', 1)
    expect(responseSuccess.body.errors).toBeUndefined()
    expect(responseSuccess.body.data).toMatchSnapshot()
  })

  it('Limit can be enforced.', async () => {
    const response = await request(server)
      .get(
        `/graphql?query=
          query getLocations {
            locations: location {
              id
            }
          }
          &operationName=getLocations`
      )
      .set('userId', 1)
    expect(response.body.errors).toBeUndefined()
    expect(response.body.data.locations.length).toBe(2)

    const responseSmaller = await request(server)
      .get(
        `/graphql?query=
          query getLocations {
            locations: location(limit: 1) {
              id
            }
          }
          &operationName=getLocations`
      )
      .set('userId', 1)
    expect(responseSmaller.body.errors).toBeUndefined()
    expect(responseSmaller.body.data.locations.length).toBe(1)

    const responseTooBig = await request(server)
      .get(
        `/graphql?query=
          query getLocations {
            locations: location(limit: 100) {
              id
            }
          }
          &operationName=getLocations`
      )
      .set('userId', 1)
    expect(responseTooBig.body.errors).toBeUndefined()
    expect(responseTooBig.body.data.locations.length).toBe(2)

    const responseNull = await request(server)
      .get(
        `/graphql?query=
          query getLocations {
            locations: location(limit: null) {
              id
            }
          }
          &operationName=getLocations`
      )
      .set('userId', 1)
    expect(responseNull.body.errors).toBeUndefined()
    expect(responseNull.body.data.locations.length).toBe(2)
  })
})
