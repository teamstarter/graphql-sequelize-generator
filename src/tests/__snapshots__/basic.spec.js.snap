// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the API queries Check that you can query a model and an association: All companies 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Company 1",
    "users": Array [
      Object {
        "id": 5,
        "name": "Mr Test 5 c 2",
      },
      Object {
        "id": 10,
        "name": "Test 10 c 2",
      },
    ],
  },
  Object {
    "id": 3,
    "name": "Company 3",
    "users": Array [],
  },
  Object {
    "id": 5,
    "name": "Company 5",
    "users": Array [],
  },
  Object {
    "id": 7,
    "name": "Company 7",
    "users": Array [],
  },
]
`;

exports[`Test the API queries Check that you can query sub associations: All companies with users and their department 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Company 1",
    "users": Array [
      Object {
        "department": Object {
          "id": 1,
        },
        "id": 5,
        "name": "Mr Test 5 c 2",
      },
      Object {
        "department": Object {
          "id": 1,
        },
        "id": 10,
        "name": "Test 10 c 2",
      },
    ],
  },
  Object {
    "id": 3,
    "name": "Company 3",
    "users": Array [],
  },
  Object {
    "id": 5,
    "name": "Company 5",
    "users": Array [],
  },
  Object {
    "id": 7,
    "name": "Company 7",
    "users": Array [],
  },
]
`;

exports[`Test the API queries Check that you can sort by virtual fields in any order: All companies sorted by userCount desc and name asc. 1`] = `
Array [
  Object {
    "id": 7,
    "name": "Company 7",
    "userCount": 70,
  },
  Object {
    "id": 5,
    "name": "Company 5",
    "userCount": 50,
  },
  Object {
    "id": 3,
    "name": "Company 3",
    "userCount": 30,
  },
  Object {
    "id": 1,
    "name": "Company 1",
    "userCount": 10,
  },
]
`;

exports[`Test the API queries Check that you can sort: All companies sorted by name 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Company 1",
  },
  Object {
    "id": 3,
    "name": "Company 3",
  },
  Object {
    "id": 5,
    "name": "Company 5",
  },
  Object {
    "id": 7,
    "name": "Company 7",
  },
]
`;

exports[`Test the API queries Limit can be enforced.: Limit and offset optimization works with multi primary key tables. 1`] = `
Object {
  "userLocations": Array [
    Object {
      "locationId": 2,
      "userId": "1",
    },
    Object {
      "locationId": 2,
      "userId": "2",
    },
    Object {
      "locationId": 2,
      "userId": "3",
    },
    Object {
      "locationId": 2,
      "userId": "4",
    },
    Object {
      "locationId": 2,
      "userId": "5",
    },
    Object {
      "locationId": 2,
      "userId": "6",
    },
    Object {
      "locationId": 2,
      "userId": "7",
    },
    Object {
      "locationId": 2,
      "userId": "8",
    },
    Object {
      "locationId": 2,
      "userId": "9",
    },
    Object {
      "locationId": 2,
      "userId": "10",
    },
  ],
}
`;

exports[`Test the API queries One can exclude a basic field from a model 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "GRAPHQL_VALIDATION_FAILED",
    },
    "locations": Array [
      Object {
        "column": 15,
        "line": 6,
      },
    ],
    "message": "Cannot query field \\"updatedAt\\" on type \\"department\\". Did you mean \\"createdAt\\"?",
  },
]
`;

exports[`Test the API queries One can exclude a model from the root of the server. 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "GRAPHQL_VALIDATION_FAILED",
    },
    "locations": Array [
      Object {
        "column": 13,
        "line": 3,
      },
    ],
    "message": "Cannot query field \\"companySetting\\" on type \\"Root_Query\\". Did you mean \\"companyType\\"?",
  },
]
`;

exports[`Test the API queries One can exclude a model from the root of the server. 2`] = `
Object {
  "company": Array [
    Object {
      "id": 1,
      "settings": Object {
        "whiteLabelEnabled": true,
      },
    },
    Object {
      "id": 3,
      "settings": null,
    },
    Object {
      "id": 5,
      "settings": null,
    },
    Object {
      "id": 7,
      "settings": null,
    },
  ],
}
`;

exports[`Test the API queries One can exclude an associated model 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "GRAPHQL_VALIDATION_FAILED",
    },
    "locations": Array [
      Object {
        "column": 15,
        "line": 6,
      },
    ],
    "message": "Cannot query field \\"company\\" on type \\"department\\". Did you mean \\"companyId\\"?",
  },
]
`;

exports[`Test the API queries Sorting can ignore spacing typoes: All companies sorted by userCount desc and name asc. 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Company 1",
    "userCount": 10,
  },
  Object {
    "id": 3,
    "name": "Company 3",
    "userCount": 30,
  },
  Object {
    "id": 5,
    "name": "Company 5",
    "userCount": 50,
  },
  Object {
    "id": 7,
    "name": "Company 7",
    "userCount": 70,
  },
]
`;

exports[`Test the API queries Sorting work on secondary field and associated fields: All companies sorted by userCount desc and name asc. 1`] = `
Array [
  Object {
    "id": 5,
    "name": "Company 5",
    "type": null,
    "userCount": 50,
  },
  Object {
    "id": 1,
    "name": "Company 1",
    "type": null,
    "userCount": 10,
  },
  Object {
    "id": 7,
    "name": "Company 7",
    "type": Object {
      "id": 2,
    },
    "userCount": 70,
  },
  Object {
    "id": 3,
    "name": "Company 3",
    "type": Object {
      "id": 2,
    },
    "userCount": 30,
  },
]
`;
